File: 2025/08 August/27 Count the number of possible triangles/main.py
# Link: https://www.geeksforgeeks.org/problems/count-possible-triangles-1587115620/1


class Solution:
    def countTriangles(self, arr):
        res = 0
        arr.sort()

        for i in range(2, len(arr)):
            left, right = 0, i - 1

            while left < right:
                if arr[left] + arr[right] > arr[i]:
                    res += right - left
                    right -= 1

                else:
                    left += 1

        return res


if __name__ == "__main__":
    sol = Solution()
    arr = [4, 6, 3, 7]
    result = sol.countTriangles(arr)
    print(f"Number of possible triangles: {result}")

    arr = [10, 21, 22, 100, 101, 200, 300]
    result = sol.countTriangles(arr)
    print(f"Number of possible triangles: {result}")

    arr = [1, 2, 3]
    result = sol.countTriangles(arr)
    print(f"Number of possible triangles: {result}")

File: 2025/08 August/27 Count the number of possible triangles/main.cpp
// Link: https://www.geeksforgeeks.org/problems/count-possible-triangles-1587115620/1

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int countTriangles(vector<int> &arr)
    {
        int res = 0;
        sort(arr.begin(), arr.end());

        for (int i = 2; i < arr.size(); ++i)
        {
            int left = 0, right = i - 1;

            while (left < right)
            {
                if (arr[left] + arr[right] > arr[i])
                {
                    res += right - left;
                    right--;
                }
                else
                {
                    left++;
                }
            }
        }

        return res;
    }
};

int main()
{
    Solution sol;
    vector<int> arr = {4, 6, 3, 7};
    int result = sol.countTriangles(arr);
    cout << "Number of possible triangles: " << result << endl;

    arr = {10, 21, 22, 100, 101, 200, 300};
    result = sol.countTriangles(arr);
    cout << "Number of possible triangles: " << result << endl;

    arr = {1, 2, 3};
    result = sol.countTriangles(arr);
    cout << "Number of possible triangles: " << result << endl;

    return 0;
}